### ISSUE TITLE
Battle Chip Gate Compatibility (via third-party tools)

### ISSUE THREAD

**Issue by DeCoded-Void** [mgba-emu/mgba#3212]
The battle chip gate accessory have already been implemented in mGBA and GBE+.

Thanks to community efforts, there are a good amount of online resources related to the accessory. [[1](https://www.therockmanexezone.com/general/2020/05/01/battle-chip-gate-homebrew-released/)] [[2](https://docs.google.com/spreadsheets/d/1_l_NRAfDAaY24yJj4kHbDwBhf0Tkt1HJ0ovubgbcvHU/edit#gid=0)] [[3](https://forums.therockmanexezone.com/dip-chip-creation-guide-t4416.html)] [[4](https://github.com/goodtofufriday/megaman-battle--network-exe-chip-project)]
It is currently easy to reproduce chips through Copper Etching on a 2mm copper clad or through a PCB fabrication service for cheap.

My current goal is to use real chips for mGBA utilizing a cheap microcontroller and a card reader. Since there aren't any information on the card-reading port of the accessory, [I 3d modeled a working one](https://github.com/mgba-emu/mgba/files/15329251/2mm.11.pin.chip.connector.zip) which uses donor pins from another edge connector.

Shonumi's implementation of the chip gate in GBE+ leverages TCP connections though their implementation of SIO with the gate, which allows for easy interaction by having a script send 3 bytes through that connection. Thanks to how its handled internally, its fairly easy to create tools such as a wireless version through a wifi-enabled microcontroller without the worry of handling the integrity of the emulated serial communication.

[Based on shonumi's blogpost](https://shonumi.github.io/articles/art16.html), my thought process for mGBA is to capture the start signal via Lua to observe the values of the addresses of serial transfers and using that to simulate the gate's transfer sequence and writing the hex of the chip id on the next transfer cycle.

Using the attached lua script [here](https://github.com/mgba-emu/mgba/files/15329159/logging.zip) and reviewing the SIO Normal Mode and SIO Multi-Player Mode documentation, I am getting these as the values:
```
SIOMULTI0 (SIODATA32_L)
1. 0x0000
2. 0xA3D0
3. 0xA3D0
4. 0xA3D0
5. 0x8FFF
6. 0x0000

SIOMULTI1 (SIODATA32_H)
1. 0xFFC6
2. 0xFFC6
3. 0xFFC6
4. 0xFFC6
5. 0xFFC6
6. 0xFFC6
7. 0xFFC6
```

I was expecting the SIODATA32_L to have the same sequence as in the  blog, however it seems like the 6th value of mine is 0x0000 while I was expecting another AxA3D0. At the same time, SIODATA32_H is indefinitely repeating the accessory's ID. The debug console shows the correct sequence of values.

I'm not too familiar with serial transfers and manipulating it though memory, so my interpretation is probably incorrect. What I was looking to do is to have the Lua script handle the transfers themselves by looping them to keep the communication alive while accepting socket data to be used in the next available transfer cycle to prevent the emulator + game thinking that the accessory is disconnected. 

I wrote this just to see if I am approaching this the wrong way, a limitation of mGBA's Lua feature, or if it is a limitation of the current gate implementation.

**Comment by DeCoded-Void** [comment]
It seems like the following addresses are updated when a chip is inserted, the values here needs to be written in a particular order while the game acknowledges the gate accessory
```
+-------+------------+
| Frame | Address    |
+-------+------------+
| 1     | 0x0200F832 |
| 2     | 0x0200F904 |
| 10    | 0x0200F900 |
| 11    | 0x0200F902 |
+--------------------+
```
Writing to those in that order makes it work.

Its a workaround but its working, I am keeping this issue open and letting you decide if you want to enhance the feature or not.

**Comment by endrift** [comment]
I'd like to add support for just sending the data via a proper function call in the scripting API at some point. It's low priority compared to other features though.

**Comment by BlaXun** [comment]
I am really interested on this. Is this now possible using the scripting API?

**Comment by DeCoded-Void** [comment]
[DeCoded-Void/DIY-Megaman-Battle-Chip-Gate 
](https://github.com/DeCoded-Void/DIY-Megaman-Battle-Chip-Gate)
I posted my repository here and added you. The implementation depends on memory rather than simulated GBA link which is what is wanted to be created by the dev. It would open up a few things such as potentially making it easier to communicate with a real gba via an emulator and easily emulator other gba accessories without having to micromanage the memory, [a method for researching such a thing can be done with a logic gate analyzer](https://www.insidegadgets.com/2018/12/09/making-the-gameboy-link-cable-wireless-packet-based/). More insight into this can also be found by a project done for pokemon gen 1/2 to gen 3 communication [here](https://www.youtube.com/watch?v=h1KKkCfzOws). So the expectation is to translate the raw data of these signals into code to be used in a newer version of the mGBA api once the feature is implemented there.

I have done a logic gate analysis on battle network 4 after I posted the issue, which will be posted to the repository whenever I am free.

**Comment by DeCoded-Void** [comment]
For more context for the readers shonumi/gbe-plus#87 has the functionality I am looking for. 

Thinking more about it, it would be a nice to have to implement it directly to [BattleChipView.cpp](https://github.com/mgba-emu/mgba/blob/master/src/platform/qt/BattleChipView.cpp) to accept incoming connections in a similar way.

Essentially, having an ip and port for the user to enter on the gui end for the emulator to listen to and handle the data by updating the relevant fields of the gui instead of using scripting to setup the connection.

I am currently trying to develop my own "chip gate" that can either be wired or wireless which is currently working with GBE+. I am using scripting on mGBA to achieve a similar effect by modifying memory values as mentioned above, however this means that multiplayer is non-functioning with that solution.

**Comment by DeCoded-Void** [comment]
Yep definitely a few ways to go about it, I was able to get it working with QTcpSocket. 

Scripting can also be used instead of this, but that is pretty much reinventing the wheel. It can be useful for certain devs that need to know what game is currently being played though. For now, it is sufficient enough for someone to create an NFC reader, webapp, real battlechips using a microcontroller, etc.

mGBA already has helpers when writing to the sio, so technically the first byte (0x80) is not required, but it is just used to indicate the start of the buffer it should read and ignore everything else.

More customization can be added to the qt window for BattleChipView such as setting a different value for uninsertion time.

I also adjusted how the insertion checkbox works since these chips are meant to be hotswapped during normal gameplay:
- When a chip is inserted and Keep Inserted is checked, it will momentarily unattach the chip according to the uninsertion time, attach the new one, and hold it in place 
- When a chip is inserted and Keep Inserted is checked off, it will attach the chip and then unattach according to the uninsertion timer.

There was an issue where a chip would not be read sometimes with the old behavior but it looks like it went away when I added that modification. Going to test this a bit more before making a PR, I have tested this by using a phone as a 'wireless chip gate' with good results.

<img width="1440" height="742" alt="Image" src="https://github.com/user-attachments/assets/b44123af-b9a4-48fb-8942-b734bbcc997d" />

**Comment by DeCoded-Void** [comment]
Sent a PR with the feature: https://github.com/mgba-emu/mgba/pull/3568

Issue can be closed after approval, it would be nice to see scripting with the chip gate and compatibility with multiplayer in the future.
